src/cirq_iqm/devices/iqm_device_metadata.py

```diff
@@ -0,0 +1,104 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""DeviceMetadata subtype for IQM devices."""
+from __future__ import annotations
+
+from typing import Iterable, Optional, Union
+
+import cirq
+from cirq import NamedQubit, devices, ops
+from cirq.contrib.routing.router import nx
+from iqm_client import QuantumArchitectureSpecification
+
+# Mapping from IQM operation names to cirq operations
+_IQM_CIRQ_OP_MAP: dict[str, tuple[Union[type[cirq.Gate], cirq.Gate, cirq.GateFamily], ...]] = {
+    # XPow and YPow kept for convenience, Cirq does not know how to decompose them into PhasedX
+    # so we would have to add those rules...
+    'phased_rx': (cirq.ops.PhasedXPowGate, cirq.ops.XPowGate, cirq.ops.YPowGate),
+    'cz': (cirq.ops.CZ,),
+    'measurement': (cirq.ops.MeasurementGate,),
+    'barrier': (),
+}
+
+
+@cirq.value.value_equality
+class IQMDeviceMetadata(devices.DeviceMetadata):
+    """Hardware metadata for IQM devices.
+
+    Args:
+        qubits: qubits that exist on the device
+        connectivity: qubit connectivity graph of the device
+        gateset: Native gateset of the device. If None, a default IQM device gateset will be used.
+    """
+
+    QUBIT_NAME_PREFIX: str = 'QB'
+    """prefix for qubit names, to be followed by their numerical index"""
+
+    def __init__(  # pylint: disable=super-init-not-called
+        self,
+        qubits: Iterable[NamedQubit],
+        connectivity: Iterable[Iterable[NamedQubit]],
+        gateset: Optional[cirq.Gateset] = None,
+    ):
+        """Construct an IQMDeviceMetadata object."""
+        nx_graph = nx.Graph()
+        for edge in connectivity:
+            edge_qubits = list(edge)
+            nx_graph.add_edge(edge_qubits[0], edge_qubits[1])
+        self._qubits_set: frozenset[NamedQubit] = frozenset(qubits)
+        self._nx_graph = nx_graph
+
+        if gateset is None:
+            # default gateset for IQM devices
+            self._gateset = cirq.Gateset(
+                ops.PhasedXPowGate, ops.XPowGate, ops.YPowGate, ops.MeasurementGate, ops.CZPowGate()
+            )
+        else:
+            self._gateset = gateset
+
+    @classmethod
+    def from_architecture(cls, architecture: QuantumArchitectureSpecification) -> IQMDeviceMetadata:
+        """Returns device metadata object created based on architecture specification"""
+        qubits = frozenset(NamedQubit(qb) for qb in architecture.qubits)
+        connectivity = tuple({NamedQubit(qb) for qb in edge} for edge in architecture.qubit_connectivity)
```

Comment:
```suggestion
        qubits = tuple(NamedQubit(qb) for qb in architecture.qubits)
        connectivity = tuple(tuple(NamedQubit(qb) for qb in edge) for edge in architecture.qubit_connectivity)
```
Since they will be just iterated over once.